pyretailscience/analysis/product_association.py

```diff
@@ -226,83 +204,162 @@ def _calc_association(  # noqa: C901 (ignore complexity) - Excluded due to min_*
         if min_uplift < 0.0:
             raise ValueError("Minimum uplift must be greater or equal to 0.")
 
-        unique_combo_df = df[[group_col, value_col]].drop_duplicates()
-        unique_combo_df[value_col] = pd.Categorical(unique_combo_df[value_col], ordered=True)
-        unique_combo_df[group_col] = pd.Categorical(unique_combo_df[group_col], ordered=True)
-
-        sparse_matrix = csc_matrix(
-            (
-                [1] * len(unique_combo_df),
-                (
-                    unique_combo_df[group_col].cat.codes,
-                    unique_combo_df[value_col].cat.codes,
-                ),
-            ),
+        if isinstance(df, pd.DataFrame):
+            df = ibis.memtable(df)
+
+        unique_transactions = df
+        total_transactions = unique_transactions.alias("t")[group_col].nunique().name("total_count")
+
+        product_occurrences = (
+            unique_transactions.group_by(value_col)
+            .aggregate(
+                occurrences=lambda t: t[group_col].nunique(),
```

Comment:
Can you remove the trailing comma here please. It's not needed
```suggestion
                occurrences=lambda t: t[group_col].nunique()
```